Source/Dafny/AST/DafnyAst.cs

```diff
@@ -2506,10 +2506,11 @@ public ArrowType(IToken tok, ArrowTypeDecl atd, List<Type> typeArgs, Type result
         // arrows are right associative, so we need parentheses around the domain type
         domainNeedsParens = true;
       } else {
-        // if the domain type consists of a single tuple type, then an extra set of parentheses is needed
+        // if the domain type consists of a single tuple type, then an extra set of parentheses is need
         // Note, we do NOT call .AsDatatype or .AsIndDatatype here, because those calls will do a NormalizeExpand().  Instead, we do the check manually.
-        var udt = typeArgs[0].Normalize() as UserDefinedType;  // note, we do Normalize(), not NormalizeExpand(), since the TypeName will use any synonym
-        if (udt != null && udt.ResolvedClass is TupleTypeDecl) {
+        var udt = typeArgs[0].Normalize() as UserDefinedType;  // note, we do Normalize(), not NormalizeExpand(), since the TypeName will use any synony
+        
+        if (udt != null && BuiltIns.IsTupleTypeName(udt.FullName))  {
```

Comment:
The line break here seems a bit extreme, or do you prefer it?

Reply:
No, I commented the original code and wrote change in a new line and removed that line leaving the break before PR. I can change it.